#Лабораторная работа № 3. Решние линейных и нелинейных систем разными методами
 Вариант №5 \\
 Завадский Никита Валерьевич

In [1]:
import numpy as np
from prettytable import PrettyTable

#1
Создать матрицу $4х4$ из случайных целых чисел из отрезка $[-8, 8]$. Методом Грама-Шмидта ортогонализовать первые два столбца матрицы. Найти $QR$ разложение матрицы методом Грама-Шмидта. Проверить методом $np.linalg$


создание матрицы

In [2]:
A1 = np.random.randint(-8,8,(4,4))
print(A1)

[[-6  5 -4  5]
 [-7 -6 -6  6]
 [-7 -5 -5 -6]
 [ 5 -4  1 -7]]


Ортогонализация первых двух столбцов

In [3]:
a1 = A1[:,0]
a2 = A1[:,1]

g1 = a1
g2 = a2 - g1 * ((a2 @ g1) / (g1 @ g1))

print("Проверка на ортогонализацию:", g1 @ g2)

E = np.zeros((4,4))
E[:,0] = g1 / np.linalg.norm(g1)
E[:,1] = g2 / np.linalg.norm(g2)
print("Ортогонализированная матрица: \n",E)

Проверка на ортогонализацию: -3.552713678800501e-15
Ортогонализированная матрица: 
 [[-0.47583095  0.60982003  0.          0.        ]
 [-0.55513611 -0.48747369  0.          0.        ]
 [-0.55513611 -0.38615563  0.          0.        ]
 [ 0.39652579 -0.49129702  0.          0.        ]]


функция нахождение $QR$ методом Грама-Шмидта

In [4]:
def QR_decay(A):
  n, m = A.shape

  Q = np.empty((n, n))
  u = np.empty((n, n))

  u[:, 0] = A[:, 0]
  Q[:, 0] = u[:, 0] / np.linalg.norm(u[:, 0])

  for i in range(1, n): ## Находим матрицу Q через Алгоритм Грамма-Шмидта

      u[:, i] = A[:, i]
      for j in range(i):
          u[:, i] -= (A[:, i] @ Q[:, j]) * Q[:, j] ## @ - матричное умножение

      Q[:, i] = u[:, i] / np.linalg.norm(u[:, i])

  R = np.zeros((n, m))
  for i in range(n):
      for j in range(i, m):
          R[i, j] = A[:, j] @ Q[:, i] ## Находим матрицу R

  return Q, R

находим $QR$ своей функцией

In [5]:
Q1, R1 = QR_decay(A1)
print("\nQ:\n", Q1)
print("\nR:\n", R1)


Q:
 [[-0.47583095  0.60982003 -0.6163621  -0.14765567]
 [-0.55513611 -0.48747369 -0.20735231  0.64124747]
 [-0.55513611 -0.38615563  0.21528437 -0.70452847]
 [ 0.39652579 -0.49129702 -0.72852963 -0.2657802 ]]

R:
 [[12.60952021  2.14123928  8.40634681 -5.15483531]
 [ 0.          9.86990853  1.92504317  5.88027089]
 [ 0.          0.          1.90461075 -0.51792314]
 [ 0.          0.          0.          9.19683874]]


Проверка QR_decay

In [6]:
Q_np, R_np = np.linalg.qr(A1)
print("\nQ:\n", Q_np)
print("\nR:\n", R_np)


Q:
 [[-0.47583095  0.60982003  0.6163621   0.14765567]
 [-0.55513611 -0.48747369  0.20735231 -0.64124747]
 [-0.55513611 -0.38615563 -0.21528437  0.70452847]
 [ 0.39652579 -0.49129702  0.72852963  0.2657802 ]]

R:
 [[12.60952021  2.14123928  8.40634681 -5.15483531]
 [ 0.          9.86990853  1.92504317  5.88027089]
 [ 0.          0.         -1.90461075  0.51792314]
 [ 0.          0.          0.         -9.19683874]]


#2
Решить систему с помощью $QR$ разложения матрицы $А$. $QR$ разложение найти методом Грама-Шмидта. Проверить полученное решение непосредственной подстановкой в исходную систему, а также методом $np.solve$
$$\begin{cases}
15.7x_1+6.6x_2-5.7x_3+11.5x_4=-2.4\\
8.8x_1-6.7x_2+5.5x_3-4.5x_4=5.6\\
6.3x_1-5.7x_2-23.4x_3+6.6x_4=7.7\\
14.3x_1+8.7x_2-15.7x_3-5.8x_4=23.4
\end{cases}$$

создание матриц

In [7]:
A2 = np.array([[15.7, 6.6, -5.7, 11.5],
               [8.8, -6.7, 5.5, -4.5],
               [6.3, -5.7, -23.4, 6.6],
               [14.3, 8.7, -15.7, -5.8]])
B2 = np.array([-2.4, 5.6, 7.7, 23.4])
print("A:\n", A2, "\nB:\n", B2)

A:
 [[ 15.7   6.6  -5.7  11.5]
 [  8.8  -6.7   5.5  -4.5]
 [  6.3  -5.7 -23.4   6.6]
 [ 14.3   8.7 -15.7  -5.8]] 
B:
 [-2.4  5.6  7.7 23.4]


нахождение $QR$ разложение матрицы $A$

In [8]:
Q2, R2 = QR_decay(A2)
print("Q:\n", Q2, "\nR:\n", R2)

Q:
 [[ 0.65869359  0.22703847  0.22835537  0.68002215]
 [ 0.36920405 -0.68130742  0.54837989 -0.3143057 ]
 [ 0.26431654 -0.55799546 -0.7641074   0.18686282]
 [ 0.59995658  0.41583026 -0.25154165 -0.63550302]] 
R:
 [[ 23.83505821   5.58672854 -17.3282564    4.178299  ]
 [  0.          12.86151096   1.48724863  -0.4177598 ]
 [  0.           0.          23.54378096  -3.42579006]
 [  0.           0.           0.          14.15384247]]


решаем систему $RX = Q^{T}B$

In [9]:
def reverse(A,B):
  n = np.size(B)
  x = np.zeros_like(B)
  for i in range(n-1, -1, -1): x[i] = (B[i] - np.sum(A[i,i+1:] * x[i+1:]))/A[i,i]
  return x

In [10]:
X2 = reverse(R2, Q2.T @ B2)
print(X2)

[ 0.46606121  0.11028446 -0.56570877 -1.188659  ]


проверяем полученное решение непосредственной подстановкой в исходную систему

In [11]:
print("AX:", A2 @ X2, "\n B:", B2)

AX: [-2.4  5.6  7.7 23.4] 
 B: [-2.4  5.6  7.7 23.4]


проверяем полученное решение методом $np.solve$

In [12]:
print("       X:", X2, "\nnp.solve:", np.linalg.solve(A2, B2))

       X: [ 0.46606121  0.11028446 -0.56570877 -1.188659  ] 
np.solve: [ 0.46606121  0.11028446 -0.56570877 -1.188659  ]


#3
Решить систему методом Зейделя с точностью до $10^{-3}$. Проверить выполнение достаточного условия сходимости. Если условие не выполняется, в программе выполнить эквивалентные реобразования системы, после этого привести к удобному для итераций виду. Оформить итерации в виде таблицы (можно пользоваться модулями PrettyTable или Pandas). Проверить полученное решение.
$$\begin{cases}
6.15x_1-1.72x_2-1.23x_3=2.15\\
0.72x_1+5.67x_2+1.18x_3=1.43\\
2.57x_1-1.34x_2-3.68x_3=1.03
\end{cases}$$


создаём матрицы

In [87]:
A3 = np.array([[6.15, -1.72, -1.23],
               [0.72, 5.67, 1.18],
               [2.57, -1.34, -3.68]])
B3 = np.array([2.15, 1.43, 1.03])
e = 1e-3
print("A:\n", A3, "\nB:\n", B3)

A:
 [[ 6.15 -1.72 -1.23]
 [ 0.72  5.67  1.18]
 [ 2.57 -1.34 -3.68]] 
B:
 [2.15 1.43 1.03]


Преобразование системы к виду $X^{k+1}=D_1X^{k+1}+D_2X^k+C$

In [88]:
n = len(B3)
C = np.array([B3[i]/A3[i][i] for i in range(n)])
D1 = np.zeros_like(A3)
D2 = np.zeros_like(A3)
for i in range(n):
  for j in range(n):
    d = -A3[i][j] / A3[i][i]
    if i > j: D1[i][j] = d
    if i < j: D2[i][j] = d

print("D1:\n", D1, "\nD2:\n", D2, "\nC:\n", C)

D1:
 [[ 0.          0.          0.        ]
 [-0.12698413  0.          0.        ]
 [ 0.69836957 -0.36413043  0.        ]] 
D2:
 [[ 0.          0.2796748   0.2       ]
 [ 0.          0.         -0.20811287]
 [ 0.          0.          0.        ]] 
C:
 [ 0.3495935   0.25220459 -0.2798913 ]


проверка сходимости метода ($max|d_{ij}$|<1)

In [99]:
D = D1 + D2
d_max = np.max(np.abs(D))
if d_max < 1: print(d_max, "< 1 => метод сходиться")
else: print(d_max, "> 1 => метод не сходиться")

0.6983695652173912 < 1 => метод сходиться


функция для итерационного вычисления

In [100]:
def iter(X_old):
  X_new = np.zeros_like(X_old)
  for i in range(n):
    X_new[i] = C[i] + D1[i] @ X_new + D2[i] @ X_old
  return X_new

решаем систему с точностью до $10^{-3}$

In [90]:
table = PrettyTable()
table.field_names = ["Iter"] + [f"x_{i+1}" for i in range(n)] + ["norma(x_k - x_k-1)"]

i = 0
norma = np.inf
x_old = np.zeros(n)

table.add_row([i] + [x[i] for i in range(len(x))] + [''])

while e < norma:
  i += 1
  x_new = iter(x_old)
  norma = np.max(np.abs(x_new - x_old))
  table.add_row([i] + [x_new[j] for j in range(3)] + [norma])
  x_old = x_new.copy()


print(table)

+------+--------------------+---------------------+----------------------+-----------------------+
| Iter |        x_1         |         x_2         |         x_3          |   norma(x_k - x_k-1)  |
+------+--------------------+---------------------+----------------------+-----------------------+
|  0   |        0.0         |         0.0         |         0.0          |                       |
|  1   | 0.3495934959349593 |  0.2078117606572891 | -0.11141643334997883 |   0.3495934959349593  |
|  2   | 0.3854299211886281 |  0.2264482977243673 | -0.09317549497170602 |  0.03583642525366881  |
|  3   | 0.3942902785806038 |  0.2215269988516011 | -0.08519569633396576 |  0.00886035739197566  |
|  4   | 0.3945098750461749 | 0.21983841475146976 | -0.08442747198323375 | 0.0016885841001313406 |
|  5   | 0.3941912655013252 | 0.21971899572826484 | -0.08460649509170359 | 0.0003186095448496906 |
+------+--------------------+---------------------+----------------------+-----------------------+


проверка решения

In [101]:
print((A3@x_new)-B3)

[ 0.0004256  -0.00021125  0.        ]


#4
Решить нелинейное уравнение методом хорд с точностью $10^{-4}$ и комбинированным методом с точностью $10^{-5}$. До решения отделить корни графическим методом, проверить выполнение условий метода.

$-2.7x^3-1.48x^2+19.23x+6.35, -4, 4$


#5
Решить систему нелинейных уравнений методом Ньютона с точностью $10^{-4}$
$$\begin{cases}
sin(x+1.5)-y+2.9=0\\
cos(y-2)+x=0
\end{cases}
$$